### ISSUE TITLE
Feature: How to sort by function in sort-classes rule?

### ISSUE THREAD

**Issue by jon9090** [azat-io/eslint-plugin-perfectionist#350]
### What rule do you want to change?

perfectionist/sort-classes

### Describe the problem

I have this class. Is it possible to order the members by function type, in this sequence: `inject`, `input`, `output`, variables, `rxState`, `injectQuery`, `injectMutation`, and then the rest?

Thanks for the help!

### Code example

```typescript
class Foo {
   api = inject(ApiService);
   foo = input(..);
   fooChanged = output(..);

   disabledStatus = true;

   state = rxState(() => {...});

   userQuery = injectQuery(() => { ... });
   clientQuery = injectQuery(() => { ... });
      
   userMutation = injectMutation(() => { ... });
   clientMutation = injectMutation(() => { ... });

   open() {....} // other functions here
}
```

### Validations

- [X] Read the [docs](https://perfectionist.dev).
- [X] Check that there isn't [already an issue](https://github.com/azat-io/eslint-plugin-perfectionist/issues) that reports the same bug to avoid creating a duplicate.

**Comment by jon9090** [comment]
@hugop95  Thank you for the quick reply!

**Comment by jon9090** [comment]
Hi @hugop95, I checked this, and something isn't working as expected.

I made an example where you can run `npm run lint` to see that there’s no change in `index.ts`. These are the options I used, as you suggested, but the sorting isn’t coming out correctly. 

For example, lint is producing this order:
```typescript
#usersApi = inject(UsersApiService);
breadcrumbs = injectBreadcrumbs();
metadata = injectQuery(() => ({
  queryKey: ['reports', 'metadata'] as const,
  queryFn: () => this.#reportApi.get(),
}));
querystring = injectQueryString({});
state = rxState<{
```

But it should be:
```typescript
#usersApi = inject(UsersApiService);
breadcrumbs = injectBreadcrumbs();
querystring = injectQueryString({});
state = rxState<{
metadata = injectQuery(() => ({
  queryKey: ['reports', 'metadata'] as const,
  queryFn: () => this.#reportApi.get(),
}));
```

It seems the sorting is not respecting the order set in the `groups` configuration:
```javascript
groups: [
  'inject',
  'input',
  'output',
  'rxState',
  'injectQuery',
  'injectMutation',
  'property',
  'unknown',
]
```

Here's a StackBlitz example for reference: https://stackblitz.com/edit/stackblitz-starters-duzwun

the fullcode:

lint:
```
 

import globals from 'globals';
import pluginJs from '@eslint/js';
import tseslint from 'typescript-eslint';
import perfectionist from 'eslint-plugin-perfectionist';

/** @type {import('eslint').Linter.Config[]} */
export default [
  { files: ['**/*.{ts}'] },
  { languageOptions: { globals: globals.browser } },
  pluginJs.configs.recommended,
  ...tseslint.configs.recommended,
  {
    rules: {
      '@typescript-eslint/no-explicit-any': 'off',
    },
  },
  {
    plugins: {
      perfectionist,
    },
    rules: {
      'perfectionist/sort-classes': [
        'error',
        {
          groups: [
            'inject',
            'input',
            'output',
            'rxState',
            'injectQuery',
            'injectMutation',
            'property',
            'unknown',
          ],
          customGroups: [
            {
              groupName: 'injectQuery',
              selector: 'property',
              elementValuePattern: 'injectQuery*',
            },
            {
              groupName: 'injectMutation',
              selector: 'property',
              elementValuePattern: 'injectMutation*',
            },
            {
              groupName: 'inject',
              selector: 'property',
              elementValuePattern: 'inject*',
            },
            {
              groupName: 'input',
              selector: 'property',
              elementValuePattern: 'input*',
            },
            {
              groupName: 'output',
              selector: 'property',
              elementValuePattern: 'output*',
            },
            {
              groupName: 'rxState',
              selector: 'property',
              elementValuePattern: 'rxState*',
            },
          ],
        },
      ],
    },
  },
];

```

code:
```
export class Foo {
  #reportApi = inject(ReportsApiService);
  #toast = inject(HotToastService);
  #usersApi = inject(UsersApiService);
  breadcrumbs = injectBreadcrumbs();
  metadata = injectQuery(() => ({
    queryKey: ['reports', 'metadata'] as const,
    queryFn: () => this.#reportApi.get(),
  }));
  querystring = injectQueryString({});
  state = rxState<{
    limit: number;
    total: number;
    isAuthenticated: boolean;
    searchQuery: any;
  }>(({ connect, set }) => {
    set({ limit: 5 });
    connect('searchQuery', this.querystring.value$ as any);
  });
  searchQuery = this.state.signal('searchQuery');
  eventsFamilyIds = this.state.computed(
    () => this.searchQuery().eventsFamilyIds || []
  );
  eventsIds = this.state.computed(() => this.searchQuery().eventsIds || []);
  eventsFamily = computed(() =>
    markSelected(this.metadata().data?.eventsFamily, [
      ...this.eventsFamilyIds(),
      ...this.eventsIds(),
    ])
  );
  groups = computed(
    () =>
      this.metadata().data?.groups?.map((g) => ({
        value: g.groupId,
        label: g.name,
      })) || []
  );
  items = computed(
    () =>
      this.metadata().data?.items?.map((i) => ({
        value: i.value,
        label: i.name,
        parent: i.parent,
      })) || []
  );
  maxDate = computed(
    () =>
      this.metadata().data?.toDate &&
      parseISO(`${this.metadata().data?.toDate}`)
  );
  form = injectForm({
    eventsFamily: this.eventsFamily,
    maxDate: this.maxDate,
    groups: this.groups,
    items: this.items,
  });
  options = injectOptions();
  previewComponent = injectDialog(this.options.previewComponent);
  reports = injectInfiniteQuery(() => ({
    queryKey: ['reports', this.searchQuery()] as const,
    initialPageParam: 0,
    getPreviousPageParam: (firstPage) => firstPage.previousId,
    getNextPageParam: (firstPage) => firstPage.nextId,
  }));
  reportSelection = inject(ReportSelection);
  translate = inject(TranslocoService);
  isAuthenticated = this.state.signal('isAuthenticated');
  favorites = injectQuery(() => ({
    enabled: this.isAuthenticated(),
    queryKey: ['favorites'] as const,
    queryFn: () => this.#usersApi.getIdsReportsFavorites(),
  }));
  toggleFavorite = injectMutation(() => ({
    mutationFn: ({
      reportId,
      isFavorite,
    }: {
      reportId: number;
      isFavorite: boolean;
    }) =>
      isFavorite
        ? this.#usersApi.add({ reportId })
        : this.#usersApi.remove({ reportId }),
    onSuccess: () => {
      this.#toast.success('success');
    },
    onError: () => this.#toast.error('error'),
  }));
  lang = toSignal(this.translate.langChanges$);
  model = computed(() => ['fromDate', 'toDate']);
  total = this.state.signal('total');
  async openGraph() {
    console.log('open');
  }
}
```

**Comment by hugop95** [comment]
@jon9090 I have taken a quick look: a lot of your properties depend on each others, and we have internal optimizations to ensure that sorting them do not cause compilation or runtime issues. This process may cause what seems to be out-of-order issues.

I'll investigate to see if things can be improved in that stage a bit deeper, but this is currently somewhat expected in files where a lot of properties are initialized while depending on others.

**Comment by hugop95** [comment]
I'll take a simple example to illustrate the process:

Take the following initial code

```ts
export class Foo {
    metadata = injectQuery();
    querystring = injectQueryString({});
    state = rxState(() => this.querystring);
}
```

The ideal order excluding properties depending on each others would be:

```ts
    state = rxState(() => this.querystring);
    metadata = injectQuery();
    querystring = injectQueryString({});
```

However, `state` depends on `queryString`, so we fix this by placing `queryString` above `state`.

```ts
    querystring = injectQueryString({});
    state = rxState(() => this.querystring);
    metadata = injectQuery();
```

`queryString` make look out of order/outside of its group, but this is a side-effect of having to place dependencies correctly.

**Comment by jon9090** [comment]
First, thank you for your help with this.

As you can see, `inject` should come first because these items are at the beginning of the group array and don’t depend on anything else. For example, `reportSelection = inject(ReportSelection);` isn’t in the `inject` section, right?

https://stackblitz.com/edit/stackblitz-starters-duzwun

`npm run lint -- --fix`

**Comment by hugop95** [comment]
> It seems the sorting is not respecting the order set in the `groups` configuration:
> 
> ```js
> groups: [
>   'inject',
>   'input',
>   'output',
>   'rxState',
>   'injectQuery',
>   'injectMutation',
>   'property',
>   'unknown',
> ]
> ```

This configuration is different from the one in your stackblitz. Let's stick to the one in stackblitz to avoid confusion.

After a quick analysis, it seems that things are working correctly. Here are some pictures explaining the situation:

All of the following members are **elements that other class members depending on** and as such (let's call them **dependency members**)
![image](https://github.com/user-attachments/assets/2d95fdf4-7352-4b88-9432-3962f831776e)

The **actual** group sorting starts from here:
![image](https://github.com/user-attachments/assets/f5c83dc1-237f-4267-bd5c-d054b296adc0)

I have shown the first couple of groups, but you get the idea.

Because so many properties depend on others, a lot of them get placed at the top, out of their original group.

## Is it possible to put a space between each groups?

Not for rule yet, but I'll see what can be done.

## I also think if it's callback then it's not dep: maxDate = computed(() => this.metadata() ... so metadata =  it's not need to be before maxDate.

Here is a counter-example where the dependency not being before yields to different results at runtime:

```ts
function computed(func: () => void) { // Simulating
    return func()
}

class Foo {
    metadata = new Date();
    maxDate = computed(() => this.metadata)
}

console.log(new Foo().maxDate) // prints the actual date ✅
```

```ts
function computed(func: () => void) { // Simulating
    return func()
}

class Foo {
    maxDate = computed(() => this.metadata)
    metadata = new Date();
}

console.log(new Foo().maxDate) // prints undefined ❌
```

**Comment by jon9090** [comment]
You're right. Since `metadata` is a callback async function, could you suggest a way to "ignore" dependency members?

**Comment by hugop95** [comment]
There is currently no way, other than sorting the file, moving the dependencies where you want, and wrapping them in `eslint-disable`-like blocks. Note that anything triggering the rule in the file again will still re-sort dependencies, so that's not a stable solution.

We could theoretically add an option `ignoredCallbackDependencies` but this would require the user to:
- Enter the function name to ignore (`computed`).
- Enter the function's parameter indexes to ignore when said index is a callback function (`0`).

That doesn't sound user-friendly, so I'm not sure if this worth exploring.

**Comment by azat-io** [comment]
Thank you for the issue!

Added in 90a9132ff93da985572d7545ab3ef1c8926d544b.
Released in [v4.0.0](https://github.com/azat-io/eslint-plugin-perfectionist/releases/tag/v4.0.0).

If you'd like, you can support the release with a retweet:
https://x.com/azat_io_en/status/1858853714238271530
